Add debugging guide (#1)
mohanvive/editor-pugins: editor-plugins.md

#### Debugging

> The VS Code Ballerina extension allows you to either run your Ballerina program (without debugging) or debug them easily by launching its debugger.

Reviewer: ```suggestion
The VS Code Ballerina language extension comes with in-built debugging capabilities and provides the same debugging experience as the conventional VS Code Debugging.
```

Author: Fixed with https://github.com/mohanvive/editor-pugins/pull/1/commits/17555bbdb13e8902e43237f54586cccebaa982a1

> >**Info:** For more information on debugging your code using VS Code, go to the [VS Code Documentation](https://code.visualstudio.com/docs/editor/debugging).

Reviewer: Can we move this to a `debugging features` section? Because its where the users might need to access the vscode documentation.

Author: Since we are not adding `debugging features` section here, I'm keeping this section as it is. WDYT?

Reviewer: @praveennadarajah Yeah but I'm just afraid that the having a reference to the vscode guide at the top would be misleading since a user might go there and refer it instead of going through our Ballerina specific guide. IMO the only context that the user might want to refer the vscode guide is, when trying to use the debugger features (not at the introduction). WDYT?

Author: As per the offline discussion, fixed with https://github.com/mohanvive/editor-pugins/pull/1/commits/6aef9a216f40f08db8f00b9ed78ef43d681ebd72

##### Starting a Debug Session

> The VS Code Ballerina extension gives you the same debugging experience as the conventional VS Code Debugger.
> We can start a quick debug session instantly to debug a ballerina program with `CodeLens`. We also can start a debug session with configurations

Reviewer: ```suggestion
We can start a quick debug session instantly to debug a ballerina program with `CodeLens`. We also can start a debug session with configurations
```

Author: Fixed with https://github.com/mohanvive/editor-pugins/pull/1/commits/17555bbdb13e8902e43237f54586cccebaa982a1

like program arguments and environment variables by adding them into the `launch.json` file.

###### Starting a Program Debug Session

Ballerina extension provides multiple options to debug your Ballerina program and the easiest way will be using our context-aware `Debug` CodeLens support provided by the extension.
The extension can detect any program entry points (main function/service definition) on the fly, and the CodeLens will appear automatically.  

Follow the steps below to start a quick debug session.

![Start_Main Quick Debug Session](/learn/images/start-quick-main-debug-session.gif)

<br/>

1. Open the folder, which includes the Ballerina program you want to debug, and open the source file in the editor.

2. Add the debug points you require by clicking in front of the line numbers of the file you want to debug.

3. Click the `Debug` CodeLens which is just above the `main()` method.

###### Starting a Test Debug Session

> The Ballerina test functions can also be debugged with the similar steps explained above.

Reviewer: ```suggestion
The Ballerina test functions can also be debugged using the codelens. The `debug` codelens will automatically appear on top of each Ballerina test function and the users are able to execute/debug only a selected test case by clicking on the corresponding codelens, as shown below. 
```

Author: Fixed with https://github.com/mohanvive/editor-pugins/pull/1/commits/17555bbdb13e8902e43237f54586cccebaa982a1

Then, you can start a program, test, or remote debug session as shown below.

>**Info:** If you launch the debug session through VS Code, the working directory will be the Ballerina package root. However, you can use remote debugging for alternative working directories.

<br/>

####### Starting a Program Debug Session

Follow the steps below to start a program debug session.

1. Select **Ballerina Debug** from the drop-down available in the upper left corner to start a program debugging session.

2. Click the **Start Debugging** icon on the upper left corner to start debugging.

    You view the output in the **DEBUG CONSOLE**.

    ![Program Debug](/learn/images/program-debug.gif)

<br/>

####### Starting a Test Debug Session

Follow the steps below to start a test debug session.

1. Select **Ballerina Test** from the drop-down available in the upper left corner to start a test debugging session.

2. Click the **Start Debugging** icon on the upper left corner to start debugging.

    You view the output in the **DEBUG CONSOLE**.

    ![Test Debug](/learn/images/test-debug.gif)

<br/>

###### Starting a Remote Debug Session

> Follow the steps below to start a remote debug session.

Reviewer: Please mention the steps to create a launch json file in here

Author: Fixed with https://github.com/mohanvive/editor-pugins/pull/1/commits/17555bbdb13e8902e43237f54586cccebaa982a1

>**Info:** If you launch the debug session through VS Code, the working directory will be the Ballerina package root. However, you can use remote debugging for alternative working directories.

<br/>

####### Starting a Program Debug Session

Follow the steps below to start a program debug session.

1. Select **Ballerina Debug** from the drop-down available in the upper left corner to start a program debugging session.

2. Click the **Start Debugging** icon on the upper left corner to start debugging.

    You view the output in the **DEBUG CONSOLE**.

    ![Program Debug](/learn/images/program-debug.gif)

<br/>

####### Starting a Test Debug Session

Follow the steps below to start a test debug session.

1. Select **Ballerina Test** from the drop-down available in the upper left corner to start a test debugging session.

2. Click the **Start Debugging** icon on the upper left corner to start debugging.

    You view the output in the **DEBUG CONSOLE**.

    ![Test Debug](/learn/images/test-debug.gif)

<br/>

###### Starting a Remote Debug Session

Start remote debugging with [Debug Session with Configurations](#debug-session-with-configurations).

> After setting the configurations, follow the steps below to start a remote debug session.

Reviewer: This is not clear enough IMO. The users might not know what we mean by `configurations`. Please mention about the launch.json creation in here as well.

Author: Shall we change the second line as follows?

> After setting the remote debug configurations like `debuggeeHost` and `debuggeePort` in `launch.json` file, follow the steps below to start a remote debug session.

Reviewer: @praveennadarajah How can the user know about the `launch.json` file if we don't mention it in here? :) Therefore the first step should be creating a `launch.json` file it its not created already.

Author: @NipunaRanasinghe I have linked the **Debug Session with Configuration** section here `[Debug Session with Configurations](#debug-session-with-configurations)`. So when the user clicks the hyperlink it will direct to the Debug Session with Configuration section. This section have all the details about how to set the configurations using the `launch.json` file. IMO we don't have to repeat the content here.

Author: As per the offline discussion, fixed with https://github.com/mohanvive/editor-pugins/pull/1/commits/6aef9a216f40f08db8f00b9ed78ef43d681ebd72

Then, you can start a program, test, or remote debug session as shown below.

>**Info:** If you launch the debug session through VS Code, the working directory will be the Ballerina package root. However, you can use remote debugging for alternative working directories.

<br/>

####### Starting a Program Debug Session

Follow the steps below to start a program debug session.

1. Select **Ballerina Debug** from the drop-down available in the upper left corner to start a program debugging session.

2. Click the **Start Debugging** icon on the upper left corner to start debugging.

    You view the output in the **DEBUG CONSOLE**.

    ![Program Debug](/learn/images/program-debug.gif)

<br/>

####### Starting a Test Debug Session

Follow the steps below to start a test debug session.

1. Select **Ballerina Test** from the drop-down available in the upper left corner to start a test debugging session.

2. Click the **Start Debugging** icon on the upper left corner to start debugging.

    You view the output in the **DEBUG CONSOLE**.

    ![Test Debug](/learn/images/test-debug.gif)

<br/>

###### Starting a Remote Debug Session

Follow the steps below to start a remote debug session.

> 1. Create the `launch.json` file to add debug configurations if it is not created already. Refer [Debug Session with Configurations](#debug-session-with-configurations) section on how to create the `launch.json` file.

Reviewer: ```suggestion
1. Create the `launch.json` configuration file if it is not created already. Refer [Debug Session with Configurations](#debug-session-with-configurations) section on how to create the `launch.json` file.
```

> 2. Add relevant `debuggeeHost` and `debuggeePort` under `Ballerina Remote` configurations section in the `launch.json` file.

Reviewer: ```suggestion
2. Open `launch.json` file and configure `debuggeeHost` and `debuggeePort` attributes under `Ballerina Remote` configurations section accordingly.
```

3. After setting the remote debug configurations, select **Ballerina Remote** from the drop-down available in the upper left corner to start a remote debugging session.

4. Open the Terminal and execute the Ballerina command, which you want to debug, out of the supported remote debugging commands below. 

    - Debugging a Ballerina package or a single file: 

    ```bash
    bal run --debug <DEBUGGEE_PORT> <BAL_FILE_PATH/PACKAGE_PATH>
    ```

   - Debugging Ballerina executable JAR:  

    ```bash 
    bal run --debug <DEBUGGEE_PORT> <EXECUTABLE_JAR_FILE_PATH>
    ```

    - Debugging Ballerina tests: 

    ```bash
    bal test --debug <DEBUGGEE_PORT> <PACKAGE_PATH>
    ```
    
    The terminal will show the following log:

    ```bash
    Listening for transport dt_socket at address: 5005
    ```

5. Click the **Start Debugging** icon on the upper left corner to start debugging.

    You view the output in the **DEBUG CONSOLE**.

    ![Remote Debug](/learn/images/remote-debug.gif)

<br/>

##### Using the Debugging Features

> The following debugging features are supported by Ballerina.

Reviewer: ```suggestion
The following debugging features are currently supported by Ballerina.
```
